Monitor app launch performance for iOS and Android in RUM (#33710)
DataDog/documentation: content/en/real_user_monitoring/application_monitoring/ios/application_launch_monitoring.md

---
title: iOS Mobile App Launch Monitoring
description: "Measure Android mobile application launch performance, including the time to initial display and time to full display."
aliases:
- /real_user_monitoring/ios/data_collected/
- /real_user_monitoring/mobile_and_tv_monitoring/application_launch_monitoring/ios/
- /real_user_monitoring/mobile_and_tv_monitoring/ios/application_launch_monitoring/
further_reading:
  - link: "https://github.com/DataDog/dd-sdk-ios"
    tag: "Source Code"
    text: "Source code for dd-sdk-ios"
  - link: "/real_user_monitoring/"
    tag: "Documentation"
    text: "Datadog Real User Monitoring"
---

<div class="alert alert-danger">

>   The <code>application_start</code> action is no longer collected starting from Android SDK version 3.5.0. The <code>rum.measure.app.startup_time metric</code> is marked as deprecated but continues to report data from devices running app versions that use older SDK versions.

Reviewer: ```suggestion
  The <code>application_start</code> action is not collected in Android SDK versions 3.5.0+. The <code>rum.measure.app.startup_time metric</code> is marked as deprecated but continues to report data from devices running app versions that use older SDK versions.
```

The time to full display is manually defined using the `Monitor.reportAppFullyDisplayed()` API in the iOS SDK based on the application’s specific definition of “fully drawn.” 

The example below calls `Monitor.reportAppFullyDisplayed()` to measure the time to full display after the app finishes loading the data needed to render its initial UI. The signal is sent only once the UI has been updated with the received content or the error has been properly handled

{{< tabs >}}
{{% tab "Swift" %}}

```swift
@MainActor
class HomeViewController: UIViewController {
    override func viewDidLoad() {
        super.viewDidLoad()
        Task { await loadData() }
    }

    private func loadData() async {
        let url = URL(string: "https://api.example.com/data")!
        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            updateUI(with: data)
        } catch {
            // Handle the error if needed
        }
        
        // Mark the app as fully displayed
        RUMMonitor.shared().reportAppFullyDisplayed()
    }
}
```
{{% /tab %}}
{{< /tabs >}}

<div class="alert alert-danger">

>   If the time to full display is not defined, the iOS SDK will only collect the time to initial display. 

Reviewer: ```suggestion
  If the time to full display is not defined, the iOS SDK only collects the time to initial display. 
```

The TTID and TTFD can be queried in the RUM Sessions Explorer using the following attributes on the vital event type:
- `@vital.type: app_launch`
- `@vital.name: time_to_initial_display` or `@vital.name: time_to_full_display`

Each TTID and TTFD side panel contains a distribution visualization, an indication of whether the launch was cold or warm, and an event waterfall. 

  {{< img src="real_user_monitoring/ios/ios-ttid-vital.png" alt="Time to initial display vital event" style="width:90%;">}}

If iOS profiling is enabled and the session is sampled, the time to initial display may include profiling data.

### Metrics
The time to initial display and time to full display are calculated as metrics:
- `rum.measure.app.startup_to_initial_display`, which represents the time to initial display 
- `rum.measure.app.startup_to_full_display`, which represents the time to full display

These metrics contain attributes to specify the launch type for accurate monitoring:
- `@vital.startup_type`
- `@vital.is_prewarmed`

<div class="alert alert-danger">
  The <code>rum.measure.app.startup_to_full_display</code> metric will not be calculated if the time to full display is not defined. 
</div>

> ## Further Reading

Reviewer: ```suggestion
## Further reading
```
